src/req_wrap-inl.h

```diff
@@ -32,11 +32,131 @@ void ReqWrap<T>::Dispatched() {
+  }
+};
+
+// Detect `void uv_foo(uv_req_t* request, ...);`.
+template <typename ReqT, typename... Args>
+struct CallLibuvFunction<ReqT, void(*)(ReqT*, Args...)> {
+  typedef void(*T)(ReqT*, Args...);
+  template <typename... PassedArgs>
+  static int Call(T fn, uv_loop_t* loop, ReqT* req, PassedArgs... args) {
+    fn(req, args...);
+    return 0;
+  }
+};
+
+// This is slightly darker magic: This template is 'applied' to each parameter
+// passed to the libuv function. If the parameter type (aka `T`) is a
+// function type, it is assumed that this it is the request callback, and a
+// wrapper that calls the original callback is created.
+// If not, the parameter is passed through verbatim.
+template <typename ReqT, typename T>
+struct MakeLibuvRequestCallback {
+  static T For(ReqWrap<ReqT>* req_wrap, T v) {
+    static_assert(!is_callable<T>::value,
+                  "MakeLibuvRequestCallback missed a callback");
+    return v;
+  }
+};
+
+// Match the `void callback(uv_req_t*, ...);` signature that all libuv
+// callbacks use.
+template <typename ReqT, typename... Args>
+struct MakeLibuvRequestCallback<ReqT, void(*)(ReqT*, Args...)> {
+  typedef void(*T)(ReqT* req, Args... args);
+
+  static void Wrapper(ReqT* req, Args... args) {
+    ReqWrap<ReqT>* req_wrap = ContainerOf(&ReqWrap<ReqT>::req_, req);
+    req_wrap->env()->DecreaseWaitingRequestCounter();
+    T original_callback = reinterpret_cast<T>(req_wrap->original_callback_);
+    original_callback(req, args...);
+  }
+
+  static T For(ReqWrap<ReqT>* req_wrap, T v) {
+    CHECK_EQ(req_wrap->original_callback_, nullptr);
+    req_wrap->original_callback_ =
+        reinterpret_cast<typename ReqWrap<ReqT>::callback_t>(v);
+    return Wrapper;
+  }
+};
+
+template <typename T>
+template <typename LibuvFunction, typename... Args>
+int ReqWrap<T>::Dispatch(LibuvFunction fn, Args... args) {
+  Dispatched();
+
+  // This expands as:
+  //
+  // int err = fn(env()->event_loop(), req(), arg1, arg2, Wrapper, arg3, ...)
+  //              ^                                       ^        ^
+  //              |                                       |        |
+  //              \-- Omitted if `fn` has no              |        |
+  //                  first `uv_loop_t*` argument         |        |
+  //                                                      |        |
+  //        A function callback whose first argument      |        |
+  //        matches the libuv request type is replaced ---/        |
+  //        by the `Wrapper` method defined above                  |
+  //                                                               |
+  //               Other (non-function) arguments are passed  -----/
+  //               through verbatim
+  int err = CallLibuvFunction<T, LibuvFunction>::Call(
+      fn,
+      env()->event_loop(),
+      req(),
+      MakeLibuvRequestCallback<T, Args>::For(this, args)...);
+  if (err >= 0)
+    env()->IncreaseWaitingRequestCounter();
+  return err;
+}
```

Comment:
Just thinking out aloud but maybe there's a way to make it more flexible / comprehensible / less prone to obscure compile errors?  Quick sketch:
```c++
template <typename T>
class LibuvCallback {
 public:
  using InitF = std::function<int(*)(T*, void (*)(T*, int))>;
  using DoneF = std::function<void(*)(std::unique_ptr<LibuvCallback>, int)>;

  T* req() { return &req_; }

 private:
  template <typename> friend int MakeLibuvCallback(Environment*, InitF, DoneF);
  std::unique_ptr<LibuvCallback> self_;
  Environment* const env_;
  DoneF done_;
  T req_;

  LibuvCallback(Environment* env, DoneF done) : env_(env), done_(done), req_() {}
  
  static void Callback(T* req, int status) {
    LibuvCallback* that = ContainerOf(&LibuvCallback::req_, req);
    that->env_->DecreaseWaitingRequestCounter();
    that->done_(std::move(that->self_), status);
  }
};

template <typename T>
int MakeLibuvCallback(
    Environment* env, LibuvCallback<T>::InitF init, LibuvCallback<T>::DoneF done) {
  using C = LibuvCallback<T>;
  std::unique_ptr<C> that(new C(env, done));
  if (int err = init(&that->req_, C::Callback)) return err;
  env->IncreaseWaitingRequestCounter();
  that->self_ = std::move(that);
  return 0;
}
```
And used like this:
```c++
{
  Environment* env = /* ... */;
  const char* path = /* ... */;
  auto init_cb = [=] (uv_fs_t* req, uv_fs_cb cb) {
    return uv_fs_stat(env->event_loop(), req, path, cb);
  };
  auto done_cb = [=] (std::unique_ptr<LibuvCallback<uv_fs_t>> that, int status) {
    // do something with that->req()
    // auto-deleted when |that| goes out of scope
  };
  int err = MakeLibuvCallback(env, init_cb, done_cb);
  // ...
}
```
The std::unique_ptr is to make it possible to extend the lifetime of the request beyond the done callback.  It shouldn't do more heap allocations than currently, assuming the compiler is smart enough to unbox the init callback.

Reply:
I like the approach, and I’ll think about whether we can implement it here, but one of the reasons for the current complexity is that libuv requests don’t always match a `(uv_foo_t* request, int status)` signature. Otherwise this would be a lot simpler. :)